aten/src/ATen/functorch/BatchedTensorImpl.cpp

```diff
@@ -103,6 +103,28 @@ const char* BatchedTensorImpl::tensorimpl_type_name() const {
   return "BatchedTensorImpl";
 }
 
+c10::intrusive_ptr<TensorImpl> BatchedTensorImpl::shallow_copy_and_detach(
+    const c10::VariableVersion& version_counter,
+    bool allow_tensor_metadata_change) const {
+  DispatchKeySet key_set = getKeysToPropagateToWrapper(value());
+  auto impl = c10::make_intrusive<BatchedTensorImpl>(key_set, value(), bdim(), level());
+  impl->set_version_counter(version_counter);
+  return impl;
+}
+
+c10::intrusive_ptr<TensorImpl> BatchedTensorImpl::shallow_copy_and_detach(
+    c10::VariableVersion&& version_counter,
+    bool allow_tensor_metadata_change) const {
+  DispatchKeySet key_set = getKeysToPropagateToWrapper(value());
+  auto impl = c10::make_intrusive<BatchedTensorImpl>(key_set, value(), bdim(), level());
+  impl->set_version_counter(version_counter);
+  return impl;
+}
+
+void BatchedTensorImpl::shallow_copy_from(const c10::intrusive_ptr<TensorImpl>& impl) {
+  TORCH_CHECK(false, "mutating directly with `.data` inside functorch transform is not allowed.");
+}
+
```

Comment:
In `shallow_copy_and_detach` of the base `TensorImpl`, we propagate the `key_set` which includes `FunctorchBatched` key. So we create a Tensor which pretends to be a batched tensor but in reality isn't, which leads to problem downstream when we try to get a field or call method exclusive to BatchedTensor.

https://github.com/pytorch/pytorch/blob/517a432d6e2feea29b08e8dca1b223c61c96ac00/c10/core/TensorImpl.cpp#L782-L796

As for the implication, I didn't understand the complete semantics of these functions (and would like to discuss offline).

As for why I thought it is ok to support `shallow_copy_from` (getter for .data) was because it worked with GradTrackingTensor, so to be consistent I thought it makes sense to allow it for BatchedTensor as well. 

Though on second thought, I think we should disallow it for both of them.
```python
import torch

def foo(x):
    y = x.data
    print(y)
    y.sum()
    return x.sum()

# torch.func.vmap(foo)(torch.randn(3, 3))
torch.func.grad(foo)(torch.randn(3, 3))

```
Output:
```
GradTrackingTensor(lvl=1, value=
    tensor([[-1.6977,  0.6374,  0.0781],
            [-0.4140,  1.5172,  0.0473],
            [ 0.8435, -0.2261,  0.0345]])
)
```

Reply:
## accessing .data under vmap

Hmm, I think we should disallow accessing .data() under vmap. Under your PR, I get the following:
```py
import torch
def f(x):
    return x.data

x = torch.randn([3], requires_grad=True)

y = torch.vmap(f)(x)
```
This returns a Tensor `y` that does require True, which is not correct -- if we did this in a while loop, it would return Tensors that do not require grad.

## setting .data under vmap
We should probably also disallow setting .data() under vmap? Under your PR I get the following:
```py
import torch
def f(x, y):
    x.data = y
    return x

x = torch.randn([3])
y = torch.randn([3], requires_grad=True)

res = torch.vmap(f)(x, y)
print(res)

RuntimeError: Batching rule not implemented for aten::_has_compatible_shallow_copy_type. We could not generate a fallback.
```
So we probably want to improve the error message some how.

## setting .data under grad

The output of your script seems fine -- what is wrong with it?

Reply:
Regarding, `.data` under grad

I had following example in mind where there seems to be conflicting semantics around `.data`. We disable directly updating `.data` (second function) but allow access to `.data` which returns a shallow copy. If we mutate the shallow copy, then `x` is also updated (and may result in silently incorrect results for `x` since that mutation doesn't go from autograd). 

`.data` allowing mutation which is opaque to autograd is same semantic as PyTorch eager but currently we have conflicting semantics under `grad`, `vjp`, etc.

```python
import torch

# This works.
def foo(x):
    y = x.data
    y.copy_(torch.zeros(3, 3))
    return (x * x).sum()

# FAILS: RuntimeError: false INTERNAL ASSERT FAILED at "aten/src/ATen/functorch/TensorWrapper.cpp":137,
# please report a bug to PyTorch. NYI
def foo(x):
    x.data = torch.zeros(3, 3)
    return (x * x).sum()

print(torch.func.grad(foo)(torch.randn(3, 3)))
```